enhancements/machine-config/mco-state-reporting.md

```diff
@@ -0,0 +1,519 @@
+
+```go
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfiguration provides information to configure an operator to manage Machine Configuration.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec is the specification of the desired behavior of the Machine Config Operator
+	// +kubebuilder:validation:Required
+	Spec MachineConfigurationSpec `json:"spec"`
+
+	// status is the most recently observed status of the Machine Config Operator
+	// +optional
+	Status MachineConfigurationStatus `json:"status"`
+}
+
+type MachineConfigurationSpec struct {
+	StaticPodOperatorSpec `json:",inline"`
+
+	// component details which part of the MCO this is coming from
+	// +kubebuilder:validation:Required
+	// +required
+	Component MCOComponent `json:"component"`
+
+	// Mode describes if we are talking about this object in cluster or during bootstrap
+	Mode MCOOperationMode `json:"mode"`
+}
+
+type MachineConfigurationStatus struct {
+	StaticPodOperatorStatus `json:",inline"`
+	// mostRecentState is the most recent state reporting for each component
+	// +listType=map
+	// +listMapKey=objectName
+	// +patchMergeKey=objectName
+	// +patchStrategy=merge
+	MostRecentState []ProgressionCondition `json:"mostRecentState" patchStrategy:"merge" patchMergeKey:"objectName"`
+	// progressionHistory contains a list of events that have happened on all objects in the MCO
+	ProgressionHistory []ProgressionHistory `json:"progressionHistory"`
+	// mostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"mostRecentError"`
```

Comment:
Who and what is updating these?

Reply:
I need to update this part, I am going to be making these metav1.Conditions similar to the machineconfignode setup, sorry for the confusion.